### DISCUSSION TITLE
大型项目用 xmake 构建的最佳实践？

### DISCUSSION THREAD

**Discussion by TOMO-CAT** [xmake-io/xmake#4828] [Q&A] [answered]
### 你在什么场景下需要该功能？

如果我需要在一个大型项目中编译 static/shared 库，需要有很多中间小 target（非库类型），要对每个子文件的编译选项单独控制，最终汇总成库文件，有对应的项目参考吗？

目前我的写法都是在根目录写一个 xmake.lua 编译一个项目的产出库，希望可以将项目中的一部分源码文件单独编译成一个 target，最后再汇总成一个静态库/动态库。

我尝试过将子 target 定义成 static 后面设置 set_policy("build.merge_archive", true) 汇总成一个总的 static target 会出现静态库丢失的问题，翻看 issue 好像是会将已经 merge 的 static target 吞掉，这样就无法被多个其他 target merge。

### 描述可能的解决方案

希望大型项目中 xmake 有一种使用方式类似于 blade/bazel build //path/to/folder:target，可以拆分成多份 xmake.lua 文件。

### 描述你认为的候选方案

我尝试过 object target，但是不改动源码的情况下，其他 object target add_deps 这个 object target 时好像不会自动传递依赖关系。

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
一种走 多个 static target ，最后 link 成一个 shared target
或者走 set_kind object，它只合并 objects
或者直接 add_files 粒度配置，`add_files("src/*.c", {defines = "XXX", cxflags = "-Dxxx"})`

**Comment by TOMO-CAT** [comment]
> 一种走 多个 static target ，最后 link 成一个 shared target 或者走 set_kind object，它只合并 objects 或者直接 add_files 粒度配置，`add_files("src/*.c", {defines = "XXX", cxflags = "-Dxxx"})`

第一种方式对静态库项目不太友好，如果一个 target 没有任何源文件的话无法添加其他的 static target。另外我们现在依赖 xmake install 工具，需要对每个不导出的 target 都配置 set_default(false)，一旦漏了就会导出到成品库。

第三种方式对于 add_syslinks 和 add_packages 隔离性不太好，后续删掉对应的 源码文件时无法和它对应的 links 和 packages 对应上。而且就我们当前的项目而言会导致这个 target 的描述过于复杂。

----------------------------------------------------
我个人偏向第二种方式，但是一个 object target add_deps 另一个 object target 时不会自动合并 objects。我感觉这一块稍微改一下源码应该能实现。不过还是一个比较妥协的操作，请问一下现在企业的大型项目接入 xmake 是如何处理这个问题的。我感觉最佳方案还是出一种新的 target kind，add_deps 时自动继承 package 和 links 以及对应的产出物。不知道这个功能有在 xmake 未来的发展计划里吗？

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
不考虑新增 kind ，目前 add_deps 原本就会 继承 package 和 links 。其他的所有需求，自己走 rules 实现。或者直接 描述域 function 去 wrap 下，按需添加到 target

[a comment by an automated account is not shown]

**Comment by TOMO-CAT** [comment]
> 不考虑新增 kind ，目前 add_deps 原本就会 继承 package 和 links 。其他的所有需求，自己走 rules 实现。或者直接 描述域 function 去 wrap 下，按需添加到 target

rules 怎么继承 object target 产出的 ,o 文件呢，在 after_build 阶段写脚本往 target:objectfiles 里添加依赖的另一个 object target 的所有 objectfiles 吗？有项目是这么操作的吗参考下

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
> > 不考虑新增 kind ，目前 add_deps 原本就会 继承 package 和 links 。其他的所有需求，自己走 rules 实现。或者直接 描述域 function 去 wrap 下，按需添加到 target
> 
> rules 怎么继承 object target 产出的 ,o 文件呢，在 after_build 阶段写脚本往 target:objectfiles 里添加依赖的另一个 object target 的所有 objectfiles 吗？有项目是这么操作的吗参考下

.o 走 add_deps 不就好了么，其他配置走 add_rules

**Reply by TOMO-CAT** [reply]
如果是一个 static target add_dpes 一个 object target 是可以继承产出的 .o 的。（这里顺带问一下 add_syslinks 为什么也需要写 public = true， 理论上自动继承是否比较合理，毕竟是链接时的选项）

![image](https://github.com/xmake-io/xmake/assets/47598093/c4e623a5-358b-4003-b50c-f5e8da458ac2)

但是如果是一个 object target add_deps 一个 object target，不会自动继承 .o 文件：

![image](https://github.com/xmake-io/xmake/assets/47598093/ebd04f8a-6394-4206-a186-2ec4c7354b11)

即使我加上 add_deps("logger", {public = true}) 都不行。

另外如果是一个没有任何 file 的 object target，好像也不能通过 add_deps 其他的 object target 来组合成一个 .o 集合 target。

[a comment by an automated account is not shown]

**Comment by TOMO-CAT** [comment]
> > > 不考虑新增 kind ，目前 add_deps 原本就会 继承 package 和 links 。其他的所有需求，自己走 rules 实现。或者直接 描述域 function 去 wrap 下，按需添加到 target
> > 
> > 
> > rules 怎么继承 object target 产出的 ,o 文件呢，在 after_build 阶段写脚本往 target:objectfiles 里添加依赖的另一个 object target 的所有 objectfiles 吗？有项目是这么操作的吗参考下
> 
> .o 走 add_deps 不就好了么，其他配置走 add_rules

如果是一个 static target add_dpes 一个 object target 是可以继承产出的 .o 的。（这里顺带问一下 add_syslinks 为什么也需要写 public = true， 理论上自动继承是否比较合理，毕竟是链接时的选项）
![image](https://github.com/xmake-io/xmake/assets/47598093/fb56f9b3-f187-432d-b5e3-f1c20597bb8a)

但是如果是一个 object target add_deps 一个 object target，不会自动继承 .o 文件：

![image](https://github.com/xmake-io/xmake/assets/47598093/014209bf-f2ff-4958-86f0-8261264e5beb)


即使我加上 add_deps("logger", {public = true}) 都不行

**Comment by TOMO-CAT** [comment, marked as answer]
https://github.com/xmake-io/xmake/issues/4887
